var/spack/repos/builtin/packages/py-arm-pyart/package.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyArmPyart(PythonPackage):
+    """Python ARM Radar Toolkit.
+
+    A growing collection of weather radar algorithms and utilities build on top
+    of the Scientific Python stack and distributed under the 3-Clause BSD
+    license. Py-ART is used by the Atmospheric Radiation Measurement (ARM)
+    Climate Research Facility for working with data from a number of
+    precipitation and cloud radars, but has been designed so that it can be
+    used by others in the radar and atmospheric communities to examine,
+    processes, and analyse data from many types of weather radars."""
+
+    homepage = "https://github.com/ARM-DOE/pyart"
+    pypi = "arm_pyart/arm_pyart-1.12.7.tar.gz"
+
+    version("1.12.7", sha256="b7b23ecef270c60b017d94603941f0c117de072a10125c5f58c0685d801f9161")
+
+    variant("cartopy", description="Plot grids on maps", default=False)
+    variant("cylp", description="Linear programming solver", default=False)
+    variant("gdal", description="Output GeoTIFFs from grid objects", default=False)
+    variant("hdf5", description="Support for HDF5 files", default=False)
+    variant("rsl", description="Use RSL library", default=False)
+    variant("wradlib", description="Calculate texture of differential phase field", default=False)
+    variant("xarray", description="Work with grid dataset", default=False)
+
+    conflicts("~hdf5", when="+wradlib")
+    conflicts("~gdal", when="+wradlib")
+    conflicts("~xarray", when="+wradlib")
+
+    depends_on("python@3.6:3.10", type=("build", "run"))
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-scm@6.2:", type="build")
+    depends_on("py-cython", type="build")
+
+    depends_on("py-pytest", type="test")
+
+    depends_on("py-numpy", type=("build", "run"))
+    depends_on("py-scipy", type=("build", "run"))
+    depends_on("py-netcdf4", type=("build", "run"))
+    depends_on("py-matplotlib", type=("build", "run"))
+    depends_on("py-pooch", type=("build", "run"))
+    depends_on("py-cftime", type=("build", "run"))
+    depends_on("py-fsspec", type=("build", "run"))
+    depends_on("py-s3fs", type=("build", "run"))
+
+    depends_on("py-cartopy", type="run", when="+cartopy")
+    depends_on("py-cylp", type="run", when="+cylp")
+    depends_on("gdal+python", type="run", when="+gdal")
+    depends_on("py-h5py", type="run", when="+hdf5")
+    depends_on("rsl", type=("build", "run"), when="+rsl")
+    depends_on("py-wradlib", type="run", when="+wradlib")
+    depends_on("py-xarray@0.21.1:", type="run", when="+xarray")
```

Comment:
xarray is always required

Reply:
The documentation is not consistent on this. In the `INSTALL.rst` file, `xarray` is listed under both "Required Dependencies" and "Optional Dependencies". I think I am going to search the `imports` statements and check the dependencies that way.

Reply:
Regardless of what the docs say, it's listed under `install_requires`, so it isn't optional.

Reply:
Yea, with the docs and the commit history flipping that, I tried to interpret what was intended. Most of the `import` statements for xarray are wrapped in 
```try:                                                                                                                                        
    import xarray as xr                                                                                                                     
    _XARRAY_AVAILABLE = True                                                                                                                
except ImportError:                                                                                                                         
    _XARRAY_AVAILABLE = False     
```
but there is one file that does a straight import. While looking at the `import` statements I discovered a couple of other issues. Please confirm if my thinking below is correct or not.

1. There is an import from `pylab`. This means that `pylab`, which is not in Spack, will need to be added as a dependency.
2. There is an import of `pandas`, which means that `py-pandas` will need to be added as a dependency.
3. There is an import of `pkg_resources`, which means that `py-setuptools` needs to also be a run dependency.
4. There is an import from `shapely`, which means `py-shapely` needs to be added as a dependency.
5. The following import
```
import StringIO
```
is not valid for Python-3 and needs to be 
```
from io import StringIO
```
Python-2 can use `import StringIO` but `from io import StringIO` also works. Given that Python-2 is mostly obsolete in Spack I am thinking of patching that to `from io import StringIO` and not differentiating between Python 2 and 3.

Reply:
Many packages have optional features that require optional dependencies. As long as you don't try to import or run that optional feature, then it doesn't really matter whether or not the imports are guarded. Of course, if importing `arm_pyart` imports every other file in the library, then it may be required.

In the case of `xarray`, since it's listed under `install_requires`, the wheel that gets installed will list it as a required dependency, and `pip check` will fail if it isn't installed. So `xarray` is definitely required.

For the other dependencies, I'll leave it to your judgement. I usually stick to `install_requires` and `extras_require` personally since I can't know all possible dependencies or how they are intended to be used.

I wouldn't worry too much about Python 2 support, I'm planning on dropping it.

Reply:
> Many packages have optional features that require optional dependencies. As long as you don't try to import or run that optional feature, then it doesn't really matter whether or not the imports are guarded.

Sorry if I was not clear, the imports for 
- `pylab`
- `pandas`
- `shapely`

are non-guarded imports.